### Retromatic - best team of all time

The point of this notebook is to show two things:

1) how to connect to our Retromatic DB via a Jupyter notebook inside of our docker-compose environment (cool)
2) to find the best possible Retromatic line-up of all time!

#### Imports

In [1]:
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import func, or_

import pandas as pd
pd.options.display.max_columns = 100
import numpy as np
import json
import random

from scipy.stats import zscore, hmean
import ipywidgets as widgets
from IPython.display import display

from app import models, schemas, crud, main

In [2]:
# Set up the DB engine

SQLALCHEMY_DATABASE_URL = "mysql+mysqldb://admin:retromatic@retromatic.cjyiqodfdde5.us-east-2.rds.amazonaws.com/retromatic"

engine = create_engine(SQLALCHEMY_DATABASE_URL, pool_size=5, max_overflow=2, pool_timeout=30, pool_recycle=1800, pool_pre_ping=True)
SessionLocal = sessionmaker(autocommit=True, autoflush=True, bind=engine)
models.Base.metadata.create_all(bind=engine)
db = SessionLocal()

#### Try it out

In [3]:
db.query(models.Fielding).first().person.nameGiven

'Francis Patterson'

In [4]:
db.query(models.Pitching).first().yearID

1871

In [5]:
main.get_batter_year_stats('zobribe01', 2012, db)

{'R': 88,
 'HR': 20,
 'RBI': 74,
 'SB': 14,
 'AVG': 0.27,
 'AVG_num': 151,
 'AVG_denom': 560,
 'OBP': 0.377,
 'OBP_num': 251,
 'OBP_denom': 666}

In [6]:
main.get_positional_eligibility('zobribe01', 2012, db)

['2B', 'OF', 'SS']

In [7]:
def name_to_playerID(name):
    searcher = main.get_searcher()
    results = searcher.search(
        main.get_parser().parse(name), limit=5, scored=True, terms=True
    )
    return [(f'{result["nameFirst"]} {result["nameLast"]}', result['playerID']) for result in results]

def playerID_to_name(playerID, db):
    name = db.query(
        models.Person.nameFirst, models.Person.nameLast
    ).filter(models.Person.playerID == playerID).first()
    return f'{name[0]} {name[1]}'

def get_pitcher_year_stats(player_id: str, year: int, db: Session):
    return sum_pitching_stints(
        db.query(models.Pitching)
            .filter(models.Pitching.playerID == player_id)
            .filter(models.Pitching.yearID == year)
            .all()
    )

In [8]:
name_to_playerID('Clayton Kershaw')

[('Clayton Kershaw', 'kershcl01')]

In [9]:
def sum_pitching_stints(stints):
    if len(stints) == 0:
        return {
            "W": 0,
            "SV": 0,
            "K": 0,
            "ERA": 0.000,
            "ERA_num": 0,
            "ERA_denom": 0,
            "WHIP": 0.000,
            "WHIP_num": 0,
            "WHIP_denom": 0
        }
    elif len(stints) > 1:
        total_W = sum([s.W for s in stints])
        total_SV = sum([s.SV for s in stints])
        total_K = sum([s.SO for s in stints])
        total_IPouts = sum([s.IPouts for s in stints])
        total_ER = sum([s.ER for s in stints])
        total_BB = sum([s.BB for s in stints]) + sum([int(s.IBB) for s in stints])
        total_H = sum([s.H for s in stints])

    else:
        s = stints[0]
        total_W = s.W
        total_SV = s.SV
        total_K = s.SO
        total_IPouts = s.IPouts
        total_ER = s.ER
        total_BB = s.BB + s.IBB
        total_H = s.H

    era_num = 9 * total_ER
    era_denom = (total_IPouts / 3)
    
    whip_num = total_BB + total_H
    whip_denom = (total_IPouts / 3)

    if era_denom == 0:
        era = 10
    else:
        era = round(era_num / era_denom, 3)

    if whip_denom == 0:
        whip = 10
    else:
        whip = round(whip_num / whip_denom, 3)

    return {
        "W": total_W,
        "SV": total_SV,
        "K": total_K,
        "ERA": era,
        "ERA_num": era_num,
        "ERA_denom": era_denom,
        "WHIP": whip,
        "WHIP_num": whip_num,
        "WHIP_denom": whip_denom
    }

In [10]:
get_pitcher_year_stats('kershcl01', 2012, db)

{'W': 14,
 'SV': 0,
 'K': 229,
 'ERA': 2.53,
 'ERA_num': 576,
 'ERA_denom': 227.66666666666666,
 'WHIP': 1.045,
 'WHIP_num': 238,
 'WHIP_denom': 227.66666666666666}

#### Gather new dataset on all eligible batter/season/positions from 1985-2015

One thing we may want to do is write the results of these queries to a new database table - why bother re-computing these things on the fly?

In [11]:
eligible_positions = db.query(
    models.Fielding.playerID,
    models.Fielding.yearID,
    models.Fielding.POS
).filter(models.Fielding.yearID >= 1961) \
 .filter(models.Fielding.yearID <= 2019) \
 .group_by(models.Fielding.POS, models.Fielding.playerID, models.Fielding.yearID) \
 .having(or_(func.sum(models.Fielding.G) > 20, func.sum(models.Fielding.GS) > 10)) \
 .order_by(models.Fielding.playerID,models.Fielding.yearID) \
 .all()

In [12]:
eligibility_df = pd.DataFrame(eligible_positions)

In [13]:
position_years = eligibility_df[eligibility_df['POS'] != 'P']

In [14]:
position_years.head()

,playerID,yearID,POS
7,aaronha01,1961,OF
8,aaronha01,1962,OF
9,aaronha01,1963,OF
10,aaronha01,1964,OF
11,aaronha01,1965,OF


In [15]:
all_batting = db.query(models.Batting.playerID,
                       models.Batting.yearID,
                       models.Batting.stint,
                       models.Batting.AB,
                       models.Batting.R,
                       models.Batting.H,
                       models.Batting.HR,
                       models.Batting.RBI,
                       models.Batting.SB,
                       models.Batting.BB,
                       models.Batting.HBP,
                       models.Batting.SF
                      ) \
    .filter(models.Batting.yearID >= 1961) \
    .filter(models.Batting.yearID <= 2019) \
    .all()

In [16]:
all_batting_df = pd.DataFrame(all_batting)

In [17]:
all_batting_df.head()

,playerID,yearID,stint,AB,R,H,HR,RBI,SB,BB,HBP,SF
0,aaronha01,1961,1,603,115,197,34,120,21,56,2,9
1,adairje01,1961,1,386,41,102,9,37,5,35,2,4
2,adcocjo01,1961,1,562,77,160,35,108,2,59,2,2
3,aguirha01,1961,1,9,0,0,0,1,0,0,0,1
4,allenbo04,1961,1,12,0,2,0,0,0,0,0,0


We have a version of `sum_stints` in `app.main` but it uses SQLAlchemy objects, which let you look up attributes with dot notation (`stint.R` instead of `stint['R']`).

In [18]:
def sum_batting_stints(stints):
    if len(stints) == 0:
        return {
            "R": 0,
            "HR": 0,
            "RBI": 0,
            "SB": 0,
            "AVG": 0.000,
            "AVG_num": 0,
            "AVG_denom": 0,
            "OBP": 0.000,
            "OBP_num": 0,
            "OBP_denom": 0,
        }
    elif len(stints) > 1:
        total_R = sum([s['R'] for s in stints])
        total_HR = sum([s['HR'] for s in stints])
        total_RBI = sum([s['RBI'] for s in stints])
        total_SB = sum([s['SB'] for s in stints])
        total_H = sum([s['H'] for s in stints])
        total_AB = sum([s['AB'] for s in stints])
        total_BB = sum([s['BB'] for s in stints])
        total_HBP = sum([s['HBP'] for s in stints])
        total_SF = sum([s['SF'] for s in stints])
    else:
        s = stints[0]
        total_R = s['R']
        total_HR = s['HR']
        total_RBI = s['RBI']
        total_SB = s['SB']
        total_H = s['H']
        total_AB = s['AB']
        total_BB = s['BB']
        total_HBP = s['HBP']
        total_SF = s['SF']

    avg_num = total_H
    avg_denom = total_AB

    obp_num = total_H + total_BB + total_HBP
    obp_denom = total_AB + total_BB + total_HBP + total_SF

    if avg_denom == 0:
        avg = 0.000
    else:
        avg = round(avg_num / avg_denom, 3)

    if obp_denom == 0:
        obp = 0.000
    else:
        obp = round(obp_num / obp_denom, 3)

    return {
        "R": total_R,
        "HR": total_HR,
        "RBI": total_RBI,
        "SB": total_SB,
        "AVG": avg,
        "AVG_num": avg_num,
        "AVG_denom": avg_denom,
        "OBP": obp,
        "OBP_num": obp_num,
        "OBP_denom": obp_denom,
    }

In [19]:
def batter_year_to_summary(batter_year):
    batter_year_stats = sum_batting_stints(
        all_batting_df[
            (all_batting_df['playerID'] == batter_year['playerID']) &
            (all_batting_df['yearID'] == batter_year['yearID'])
        ].to_dict('records')
    )
    return pd.Series(batter_year_stats)

In [20]:
def sum_pitching_stints(stints):
    if len(stints) == 0:
        return {
            "W": 0,
            "SV": 0,
            "K": 0,
            "ERA": 0.000,
            "ERA_num": 0,
            "ERA_denom": 0,
            "WHIP": 0.000,
            "WHIP_num": 0,
            "WHIP_denom": 0
        }
    elif len(stints) > 1:
        total_W = sum([s['W'] for s in stints])
        total_SV = sum([s['SV'] for s in stints])
        total_K = sum([s['SO'] for s in stints])
        total_IPouts = sum([s['IPouts'] for s in stints])
        total_ER = sum([s['ER'] for s in stints])
        total_BB = sum([s['BB'] for s in stints]) + sum([int(s['IBB']) for s in stints])
        total_H = sum([s['H'] for s in stints])

    else:
        s = stints[0]
        total_W = s['W']
        total_SV = s['SV']
        total_K = s['SO']
        total_IPouts = s['IPouts']
        total_ER = s['ER']
        total_BB = s['BB'] + s['IBB']
        total_H = s['H']

    era_num = 9 * total_ER
    era_denom = (total_IPouts / 3)
    
    whip_num = total_BB + total_H
    whip_denom = (total_IPouts / 3)

    if era_denom == 0:
        era = 10
    else:
        era = round(era_num / era_denom, 3)

    if whip_denom == 0:
        whip = 10
    else:
        whip = round(whip_num / whip_denom, 3)

    return {
        "W": total_W,
        "SV": total_SV,
        "K": total_K,
        "ERA": era,
        "ERA_num": era_num,
        "ERA_denom": era_denom,
        "WHIP": whip,
        "WHIP_num": whip_num,
        "WHIP_denom": whip_denom
    }

In [21]:
batter_years = position_years[['playerID','yearID']].drop_duplicates()

In [22]:
batter_years.head()

,playerID,yearID
7,aaronha01,1961
8,aaronha01,1962
9,aaronha01,1963
10,aaronha01,1964
11,aaronha01,1965


In [23]:
batter_years[['R', 'HR', 'RBI', 'SB', 'AVG', 'AVG_num', 'AVG_denom', 'OBP', 'OBP_num',
       'OBP_denom']] = batter_years.apply(batter_year_to_summary, axis=1)

In [24]:
position_years = pd.merge(
    position_years, batter_years, 
    how='left', left_on=['playerID','yearID'], right_on = ['playerID','yearID']
)

In [25]:
position_years[position_years['playerID'] == 'zobribe01']

,playerID,yearID,POS,R,HR,RBI,SB,AVG,AVG_num,AVG_denom,OBP,OBP_num,OBP_denom
28668,zobribe01,2006,SS,10.0,2.0,18.0,2.0,0.224,41.0,183.0,0.260,51.0,196.0
28669,zobribe01,2007,SS,8.0,1.0,9.0,2.0,0.155,15.0,97.0,0.184,19.0,103.0
28670,zobribe01,2008,SS,32.0,12.0,30.0,3.0,0.253,50.0,198.0,0.339,77.0,227.0
28671,zobribe01,2008,OF,32.0,12.0,30.0,3.0,0.253,50.0,198.0,0.339,77.0,227.0
28672,zobribe01,2009,OF,91.0,27.0,91.0,17.0,0.297,149.0,501.0,0.405,242.0,598.0
28673,zobribe01,2009,2B,91.0,27.0,91.0,17.0,0.297,149.0,501.0,0.405,242.0,598.0
28674,zobribe01,2010,2B,77.0,10.0,75.0,24.0,0.238,129.0,541.0,0.346,224.0,648.0
28675,zobribe01,2010,OF,77.0,10.0,75.0,24.0,0.238,129.0,541.0,0.346,224.0,648.0
28676,zobribe01,2011,OF,99.0,20.0,91.0,19.0,0.269,158.0,588.0,0.353,237.0,672.0
28677,zobribe01,2011,2B,99.0,20.0,91.0,19.0,0.269,158.0,588.0,0.353,237.0,672.0


Ok! so we have `position_years` which contains one row for each eligible player/year/position combo, where that row represents _all_ of their total stats for that year. Now...let's find the best years.

I think this is what we'd possibly write to a new DB table?

Actually yeah, let's just do that.

Let's get pitching together...

In [26]:
pitcher_years = db.query(models.Pitching.playerID,
                       models.Pitching.yearID,
                       func.sum(models.Pitching.GS).label('GS'),
                       func.sum(models.Pitching.G).label('G')
                      ) \
    .filter(models.Pitching.yearID >= 1961) \
    .filter(models.Pitching.yearID <= 2019) \
    .group_by(models.Pitching.playerID, models.Pitching.yearID).all()

In [27]:
pitcher_years = pd.DataFrame(pitcher_years)

In [28]:
def games_to_pitcher_eligibility(pitcher_year):
    POS = []
    if pitcher_year['GS'] >= 10:
        POS.append('SP')
    if (pitcher_year['G'] - pitcher_year['GS']) >= 15:
        POS.append('RP')
    
    return POS

In [29]:
pitcher_years['POS'] = pitcher_years.apply(games_to_pitcher_eligibility, axis=1)

In [30]:
pitcher_years = pitcher_years.explode('POS')

In [31]:
pitcher_years.drop(
    pitcher_years[
        (pitcher_years['POS'] != 'SP') &
        (pitcher_years['POS'] != 'RP')
    ].index
, inplace=True)                   

In [32]:
pitcher_years = pitcher_years[['playerID','yearID','POS']]

In [33]:
pitcher_years = pitcher_years.drop_duplicates()

In [34]:
all_pitching = db.query(models.Pitching.playerID,
                       models.Pitching.yearID,
                       models.Pitching.stint,
                       models.Pitching.W,
                       models.Pitching.SV,
                       models.Pitching.SO,
                       models.Pitching.ER,
                       models.Pitching.BB,
                       models.Pitching.IBB,
                       models.Pitching.H,
                       models.Pitching.IPouts
                      ) \
    .filter(models.Pitching.yearID >= 1961) \
    .filter(models.Pitching.yearID <= 2019) \
    .all()

In [35]:
all_pitching_df = pd.DataFrame(all_pitching)

In [36]:
all_pitching_df.head(10)

,playerID,yearID,stint,W,SV,SO,ER,BB,IBB,H,IPouts
0,aguirha01,1961,1,4,8,32,20,38,3,44,166
1,allenbo04,1961,1,3,3,42,34,40,5,96,245
2,anderbo01,1961,1,7,8,96,72,56,11,162,456
3,andercr01,1961,1,4,1,21,14,12,0,38,116
4,antonjo02,1961,1,0,0,23,35,18,0,68,144
5,antonjo02,1961,2,1,0,8,9,3,0,16,32
6,archeji02,1961,1,9,5,110,73,60,6,204,616
7,arrigge01,1961,1,0,0,6,11,10,0,9,29
8,arroylu01,1961,1,15,29,87,29,49,8,83,357
9,baldsja01,1961,1,5,3,59,43,49,11,90,299


In [37]:
def pitcher_year_to_summary(pitcher_year):
    pitcher_year_stats = sum_pitching_stints(
        all_pitching_df[
            (all_pitching_df['playerID'] == pitcher_year['playerID']) &
            (all_pitching_df['yearID'] == pitcher_year['yearID'])
        ].to_dict('records')
    )
    return pd.Series(pitcher_year_stats)

In [38]:
pitcher_years[['W', 'SV', 'K', 'ERA', 'ERA_num', 'ERA_denom', 'WHIP', 'WHIP_num', 'WHIP_denom']] = pitcher_years.apply(pitcher_year_to_summary, axis=1)

In [39]:
pitcher_years.head()

,playerID,yearID,POS,W,SV,K,ERA,ERA_num,ERA_denom,WHIP,WHIP_num,WHIP_denom
1,aardsda01,2006,RP,3.0,0.0,49.0,4.075,216.0,53.000000,1.302,69.0,53.000000
2,aardsda01,2007,RP,2.0,0.0,36.0,6.402,207.0,32.333333,1.825,59.0,32.333333
3,aardsda01,2008,RP,4.0,0.0,49.0,5.548,270.0,48.666667,1.767,86.0,48.666667
4,aardsda01,2009,RP,3.0,38.0,80.0,2.523,180.0,71.333333,1.206,86.0,71.333333
5,aardsda01,2010,RP,0.0,31.0,49.0,3.443,171.0,49.666667,1.268,63.0,49.666667


#### Z scores galore

So Z scores are a useful way to normalize our data and find players who truly stand out, but it is sensitive to the mean of any given category. We have tons and tons of irrelevant seasons in our dataset, where it's very unlikely those players could be fantasy relevant. So let's prune them out - if you didn't have 300 at-bats, or hit 20 HR or hit 20 steals, what are you doing in our data? Let's drop those kinds of rows.

On the pitcher side of things we'll drop pitchings with fewer than 40 innings pitched, fewer than 5 wins and fewer than 10 saves.

In [40]:
position_years.drop(
    position_years[
        (position_years['AVG_denom'] < 250) &
        (position_years['SB'] < 15) &
        (position_years['HR'] < 15)
    ].index
, inplace=True)                   

In [41]:
pitcher_years.drop(
    pitcher_years[
        (pitcher_years['ERA_denom'] < 80) &
        (pitcher_years['W'] < 5) &
        (pitcher_years['SV'] < 15)
    ].index
, inplace=True)                   

When we compute our Z scores, we want to calculate two versions: overall Z score within our player database, and a "within position" Z score within a player's positional eligibility.

This let's us ask literally who is the most valuable across the entire dataset, but also easily compare the best second-basemen seasons to the best outfield seasons based on their relative value.

In [42]:
position_years['Outs'] = position_years['AVG_denom'] - position_years['AVG_num']

In [43]:
# overall Z scores

position_years['R_Z'] =  zscore(position_years.R)
position_years['RBI_Z'] =  zscore(position_years.RBI)
position_years['HR_Z'] =  zscore(position_years.HR)
position_years['SB_Z'] =  zscore(position_years.SB)
position_years['Hits_Z'] =  zscore(position_years.AVG_num)
position_years['Outs_Z'] =  zscore(position_years.Outs)

# positional Z scores

position_years['R_POS_Z'] =  position_years.groupby('POS').R.transform(lambda x : zscore(x,ddof=1))
position_years['RBI_POS_Z'] =  position_years.groupby('POS').RBI.transform(lambda x : zscore(x,ddof=1))
position_years['HR_POS_Z'] =  position_years.groupby('POS').HR.transform(lambda x : zscore(x,ddof=1))
position_years['SB_POS_Z'] =  position_years.groupby('POS').SB.transform(lambda x : zscore(x,ddof=1))
position_years['Hits_POS_Z'] =  position_years.groupby('POS').AVG_num.transform(lambda x : zscore(x,ddof=1))
position_years['Outs_POS_Z'] =  position_years.groupby('POS').Outs.transform(lambda x : zscore(x,ddof=1))

# Total Z
position_years['Total_Z'] = position_years['R_Z'] + \
position_years['RBI_Z'] + \
position_years['HR_Z'] + \
position_years['SB_Z'] + \
position_years['Hits_Z'] - \
position_years['Outs_Z']

# POS Z
position_years['Total_POS_Z'] = position_years['R_POS_Z'] + \
position_years['RBI_POS_Z'] + \
position_years['HR_POS_Z'] + \
position_years['SB_POS_Z'] + \
position_years['Hits_POS_Z'] - \
position_years['Outs_POS_Z']

Now this would easily tell you who is best to draft if you're targeting one category, but we aren't I'm sure there are other ways to tally up across categories, but I'm going to naively sum the Z score for our 4 counting categories, then add the Z score for Hits, and subtract the Z score for Outs. I think this is cleaner than adding the Z score for average because I want to make sure someone with a .333 average on 600 at-bats gets more credit than one in 300 at-bats.

In [44]:
pitcher_years['W_LOG'] = np.log(pitcher_years['W'] + .00000000001)
pitcher_years['SV_LOG'] = np.log(pitcher_years['SV'] + .00000000001)

pitcher_years['W_Z'] =  zscore(pitcher_years.W_LOG)
pitcher_years['SV_Z'] =  zscore(pitcher_years.SV)
pitcher_years['K_Z'] =  zscore(pitcher_years.K)
pitcher_years['ERA_Z'] =  zscore(pitcher_years.ERA)
pitcher_years['WHIP_Z'] =  zscore(pitcher_years.WHIP)
pitcher_years['Outs_Z'] =  zscore(pitcher_years.ERA_denom)

pitcher_years['W_POS_Z'] =  pitcher_years.groupby('POS').W_LOG.transform(lambda x : zscore(x,ddof=1))
pitcher_years['SV_POS_Z'] =  pitcher_years.groupby('POS').SV.transform(lambda x : zscore(x,ddof=1))
pitcher_years['K_POS_Z'] =  pitcher_years.groupby('POS').K.transform(lambda x : zscore(x,ddof=1))
pitcher_years['ERA_POS_Z'] =  pitcher_years.groupby('POS').ERA.transform(lambda x : zscore(x,ddof=1))
pitcher_years['WHIP_POS_Z'] =  pitcher_years.groupby('POS').WHIP.transform(lambda x : zscore(x,ddof=1))
pitcher_years['Outs_POS_Z'] =  pitcher_years.groupby('POS').ERA_denom.transform(lambda x : zscore(x,ddof=1))

pitcher_years['Total_Z'] = .7 * (pitcher_years['W_Z'] + pitcher_years['SV_Z'] + pitcher_years['K_Z'] \
- 2 * pitcher_years['ERA_Z'] - 2 * pitcher_years['WHIP_Z'] \
+ pitcher_years['Outs_Z'])

pitcher_years['Total_POS_Z'] = (pitcher_years['Total_Z'] + .7 * (pitcher_years['W_POS_Z'] + pitcher_years['SV_POS_Z'] + pitcher_years['K_POS_Z'] \
- 2 * pitcher_years['ERA_POS_Z'] - 2 * pitcher_years['WHIP_POS_Z'] \
+ pitcher_years['Outs_POS_Z'])) / 2

Ok - so if we look at `Total_Z` vs. `Total_POS_Z` we should see that the overall list is dominated by outfielders at the top, with position player seasons typically much lower, but using the positional Z scores is probably a better ranked list of who you would actually want to draft first to get maximum value.

In [45]:
position_years.sort_values('Total_Z', ascending=False, inplace=True)
position_years.head(20)

,playerID,yearID,POS,R,HR,RBI,SB,AVG,AVG_num,AVG_denom,OBP,OBP_num,OBP_denom,Outs,R_Z,RBI_Z,HR_Z,SB_Z,Hits_Z,Outs_Z,R_POS_Z,RBI_POS_Z,HR_POS_Z,SB_POS_Z,Hits_POS_Z,Outs_POS_Z,Total_Z,Total_POS_Z
27164,walkela01,1997,OF,143.0,49.0,130.0,33.0,0.366,208.0,568.0,0.452,300.0,664.0,360.0,3.555966,2.884900,3.655069,2.144010,2.381195,0.483962,3.345387,2.686922,3.391749,1.570544,2.268135,0.474226,14.137178,12.788511
24524,sosasa01,1998,OF,134.0,66.0,158.0,18.0,0.308,198.0,643.0,0.377,272.0,722.0,445.0,3.171375,3.981440,5.379201,0.815575,2.115492,1.540624,2.963441,3.740435,5.058869,0.442883,2.007783,1.507237,13.922459,12.706174
22375,rodrial01,2007,3B,143.0,54.0,156.0,24.0,0.314,183.0,583.0,0.422,299.0,708.0,400.0,3.555966,3.903116,4.162166,1.346949,1.716937,0.981215,3.669584,3.823278,4.153101,2.432645,1.731781,0.920936,13.703919,14.889453
2639,bondsba01,2001,OF,129.0,73.0,137.0,13.0,0.328,156.0,476.0,0.515,342.0,664.0,320.0,2.957714,3.159035,6.089138,0.372763,0.999539,-0.013291,2.751248,2.950300,5.745331,0.066996,0.914302,-0.011897,13.591481,12.440075
24527,sosasa01,2001,OF,146.0,64.0,160.0,0.0,0.328,189.0,577.0,0.437,311.0,711.0,388.0,3.684162,4.059764,5.176362,-0.778547,1.876359,0.832039,3.472703,3.815686,4.862737,-0.910309,1.773465,0.814512,13.186062,12.199770
3759,burksel01,1996,OF,142.0,40.0,128.0,32.0,0.344,211.0,613.0,0.408,278.0,682.0,402.0,3.513233,2.806576,2.742293,2.055447,2.460906,1.006077,3.302949,2.611671,2.509155,1.495366,2.346241,0.984655,12.572378,11.280728
11529,henderi01,1982,OF,119.0,10.0,51.0,130.0,0.267,143.0,536.0,0.398,261.0,656.0,393.0,2.530392,-0.208908,-0.300294,10.734554,0.654125,0.894195,2.326863,-0.285489,-0.432822,8.862748,0.575844,0.875277,12.515674,10.171867
11532,henderi01,1985,OF,146.0,24.0,72.0,80.0,0.314,172.0,547.0,0.419,274.0,654.0,375.0,3.684162,0.613497,1.119580,6.306439,1.424664,0.670432,3.472703,0.504645,0.940101,5.103880,1.330866,0.656522,12.477910,10.695673
22366,rodrial01,1998,SS,123.0,42.0,124.0,46.0,0.310,213.0,686.0,0.360,268.0,745.0,473.0,2.701321,2.649927,2.945132,3.295320,2.514047,1.888701,2.862797,3.712216,4.434133,3.033419,2.501155,1.669483,12.217045,14.874236
17179,mcgwima01,1998,1B,130.0,70.0,147.0,1.0,0.299,152.0,509.0,0.470,320.0,681.0,357.0,3.000447,3.550656,5.784879,-0.689984,0.893258,0.446668,2.965924,2.952682,4.992937,-0.543396,0.766725,0.376005,12.092588,10.758866


In [46]:
position_years.sort_values('Total_POS_Z', ascending=False, inplace=True)
position_years.head(10)

,playerID,yearID,POS,R,HR,RBI,SB,AVG,AVG_num,AVG_denom,OBP,OBP_num,OBP_denom,Outs,R_Z,RBI_Z,HR_Z,SB_Z,Hits_Z,Outs_Z,R_POS_Z,RBI_POS_Z,HR_POS_Z,SB_POS_Z,Hits_POS_Z,Outs_POS_Z,Total_Z,Total_POS_Z
22428,rodriiv01,1999,C,116.0,35.0,113.0,25.0,0.332,199.0,600.0,0.356,224.0,630.0,401.0,2.402195,2.219144,2.235195,1.435511,2.142062,0.993646,4.156318,3.109037,3.205023,6.552570,3.273639,1.889261,9.440462,18.407325
22369,rodrial01,2001,SS,133.0,52.0,135.0,18.0,0.318,201.0,632.0,0.399,292.0,732.0,431.0,3.128643,3.080711,3.959327,0.815575,2.195203,1.366586,3.297363,4.236547,5.734693,0.626347,2.184749,1.168135,11.812874,14.911563
22375,rodrial01,2007,3B,143.0,54.0,156.0,24.0,0.314,183.0,583.0,0.422,299.0,708.0,400.0,3.555966,3.903116,4.162166,1.346949,1.716937,0.981215,3.669584,3.823278,4.153101,2.432645,1.731781,0.920936,13.703919,14.889453
22366,rodrial01,1998,SS,123.0,42.0,124.0,46.0,0.310,213.0,686.0,0.360,268.0,745.0,473.0,2.701321,2.649927,2.945132,3.295320,2.514047,1.888701,2.862797,3.712216,4.434133,3.033419,2.501155,1.669483,12.217045,14.874236
22370,rodrial01,2002,SS,125.0,57.0,142.0,9.0,0.300,187.0,624.0,0.392,284.0,725.0,437.0,2.786785,3.354846,4.466425,0.018514,1.823219,1.441174,2.949710,4.570212,6.384972,-0.147354,1.815608,1.239757,11.008615,14.333392
20621,piazzmi01,1997,C,104.0,40.0,124.0,5.0,0.362,201.0,556.0,0.431,273.0,633.0,355.0,1.889408,2.649927,2.742293,-0.335735,2.195203,0.421805,3.464051,3.655564,3.882526,0.795242,3.339470,1.149770,8.719291,13.987084
1922,benchjo01,1970,C,97.0,45.0,148.0,5.0,0.293,177.0,605.0,0.345,231.0,670.0,428.0,1.590283,3.589819,3.249390,-0.335735,1.557516,1.329292,3.060229,4.847988,4.560030,0.795242,2.549492,2.323310,8.321980,13.489671
22364,rodrial01,1996,SS,141.0,36.0,123.0,15.0,0.358,215.0,601.0,0.414,278.0,671.0,386.0,3.470501,2.610765,2.336614,0.549888,2.567187,0.807176,3.645016,3.664549,3.653798,0.368447,2.553889,0.630977,10.727780,13.254721
18190,morgajo02,1976,2B,113.0,27.0,111.0,60.0,0.320,151.0,472.0,0.444,266.0,599.0,321.0,2.273998,2.140820,1.423838,4.535192,0.866687,-0.000860,2.344281,3.225045,2.578104,4.113849,0.874159,-0.022024,11.241396,13.157461
22368,rodrial01,2000,SS,134.0,41.0,132.0,15.0,0.316,175.0,554.0,0.420,282.0,672.0,379.0,3.171375,2.963224,2.843712,0.549888,1.504375,0.720157,3.340819,4.093547,4.304077,0.368447,1.499202,0.547419,10.312418,13.058674


In [47]:
pitcher_years.sort_values('Total_POS_Z', ascending=False, inplace=True)
pitcher_years.head(25)

,playerID,yearID,POS,W,SV,K,ERA,ERA_num,ERA_denom,WHIP,WHIP_num,WHIP_denom,W_LOG,SV_LOG,W_Z,SV_Z,K_Z,ERA_Z,WHIP_Z,Outs_Z,W_POS_Z,SV_POS_Z,K_POS_Z,ERA_POS_Z,WHIP_POS_Z,Outs_POS_Z,Total_Z,Total_POS_Z
13713,koufasa01,1965,SP,26.0,2.0,382.0,2.038,684.0,335.666667,0.867,291.0,335.666667,3.258097,6.931472e-01,0.679393,-0.194370,5.951050,-1.751690,-2.613224,3.285627,1.587776,1.653995,5.340909,-2.084863,-2.847601,3.137281,12.916070,14.012746
8984,gibsobo01,1968,SP,22.0,0.0,268.0,1.123,342.0,304.666667,0.873,266.0,304.666667,3.091042,-2.532844e+01,0.600598,-0.424509,3.580008,-2.622730,-2.581381,2.764942,1.344034,-0.282373,3.080286,-3.020334,-2.813276,2.555355,11.850481,12.352823
13711,koufasa01,1963,SP,25.0,0.0,306.0,1.881,585.0,311.000000,0.897,279.0,311.000000,3.218876,-2.532844e+01,0.660894,-0.424509,4.370355,-1.901147,-2.454005,2.871318,1.530551,-0.282373,3.833827,-2.245375,-2.675978,2.674243,11.331854,11.825561
8550,gagneer01,2003,RP,2.0,55.0,137.0,1.202,99.0,82.333333,0.717,59.0,82.333333,0.693147,4.007333e+00,-0.530433,5.904298,0.855389,-2.547525,-3.409321,-0.969436,-0.155297,3.707347,2.946631,-2.153903,-3.018784,-0.272015,12.021457,11.810942
16594,mclaide01,1968,SP,31.0,0.0,280.0,1.955,657.0,336.000000,0.911,306.0,336.000000,3.433987,-2.532844e+01,0.762357,-0.424509,3.829591,-1.830702,-2.379703,3.291226,1.844412,-0.282373,3.318246,-2.169720,-2.595887,3.143539,11.115632,11.702079
13714,koufasa01,1966,SP,27.0,0.0,317.0,1.728,558.0,323.000000,0.997,322.0,323.000000,3.295837,-2.532844e+01,0.697194,-0.424509,4.599140,-2.046796,-1.923274,3.072874,1.642842,-0.282373,4.051957,-2.401798,-2.103901,2.899505,11.119388,11.622859
15891,martipe02,2000,SP,18.0,0.0,284.0,1.742,378.0,217.000000,0.737,160.0,217.000000,2.890372,-2.532844e+01,0.505946,-0.424509,3.912786,-2.033469,-3.303175,1.292466,1.051243,-0.282373,3.397566,-2.387485,-3.591301,0.909694,11.171983,11.547787
7055,eckerde01,1990,RP,4.0,48.0,73.0,0.614,45.0,73.333333,0.627,46.0,73.333333,1.386294,3.871201e+00,-0.203492,5.098814,-0.475722,-3.107275,-3.886979,-1.120603,0.054647,3.135831,0.264175,-2.697439,-3.447975,-0.569796,12.101253,11.362116
2293,bluevi01,1971,SP,24.0,0.0,301.0,1.817,567.0,312.000000,0.962,300.0,312.000000,3.178054,-2.532844e+01,0.641639,-0.424509,4.266362,-1.962072,-2.109030,2.888115,1.470989,-0.282373,3.734677,-2.310807,-2.304128,2.693015,10.859668,11.325996
4030,carltst01,1972,SP,27.0,0.0,310.0,1.975,684.0,346.333333,1.016,352.0,346.333333,3.295837,-2.532844e+01,0.697194,-0.424509,4.453550,-1.811663,-1.822435,3.464788,1.642842,-0.282373,3.913147,-2.149272,-1.995206,3.337514,10.821454,11.325757


Said another way, if we look at the average total Z score within each position group, we see basically the list any fantasy player would cook up if they were asked for the general impact of each position: OF, 1B, **big gap**, 3B, 2B, SS, _gap_, C.

In [48]:
position_years[['POS','Total_Z']].groupby('POS').mean().sort_values('Total_Z', ascending=False)

,Total_Z
POS,
OF,0.784401
1B,0.775007
3B,-0.100390
2B,-0.784252
SS,-1.012530
C,-1.685277


Whereas doing the same for the positional Z scores will be definition give us very close to 0 for every position.

In [49]:
position_years[['POS','Total_POS_Z']].groupby('POS').mean().sort_values('Total_POS_Z')

,Total_POS_Z
POS,
OF,-3.350256e-15
SS,-6.380110e-17
C,-4.078644e-17
2B,6.946351e-16
1B,6.978050e-16
3B,9.434363e-16


In [50]:
pitcher_years[['POS','Total_Z']].groupby('POS').mean().sort_values('Total_Z', ascending=False)

,Total_Z
POS,
SP,0.116646
RP,-0.202660


Ok, moment of truth! Who are the top 5 player-seasons at each position?

In [51]:
position_years.groupby('POS').head(5).sort_values(['POS','Total_POS_Z'],ascending=[True,False])

,playerID,yearID,POS,R,HR,RBI,SB,AVG,AVG_num,AVG_denom,OBP,OBP_num,OBP_denom,Outs,R_Z,RBI_Z,HR_Z,SB_Z,Hits_Z,Outs_Z,R_POS_Z,RBI_POS_Z,HR_POS_Z,SB_POS_Z,Hits_POS_Z,Outs_POS_Z,Total_Z,Total_POS_Z
1136,bagweje01,1999,1B,143.0,42.0,126.0,30.0,0.304,171.0,562.0,0.454,331.0,729.0,391.0,3.555966,2.728252,2.945132,1.878323,1.398094,0.869333,3.537521,2.153163,2.303949,4.511768,1.280065,0.826303,11.636433,12.960163
1134,bagweje01,1997,1B,109.0,43.0,135.0,31.0,0.286,162.0,566.0,0.425,305.0,717.0,404.0,2.103069,3.080711,3.046551,1.966885,1.158961,1.030940,2.042573,2.495814,2.399984,4.686084,1.036904,0.998476,10.325238,11.662884
8931,galaran01,1996,1B,119.0,47.0,150.0,18.0,0.304,190.0,626.0,0.357,247.0,691.0,436.0,2.530392,3.668143,3.452230,0.815575,1.902930,1.428742,2.482264,3.066899,2.784125,2.419976,1.793405,1.422286,10.940526,11.124383
21127,pujolal01,2009,1B,124.0,47.0,135.0,16.0,0.327,186.0,568.0,0.443,310.0,700.0,382.0,2.744053,3.080711,3.452230,0.638450,1.796648,0.757451,2.702109,2.495814,2.784125,2.071344,1.685333,0.707107,10.954641,11.031619
17179,mcgwima01,1998,1B,130.0,70.0,147.0,1.0,0.299,152.0,509.0,0.470,320.0,681.0,357.0,3.000447,3.550656,5.784879,-0.689984,0.893258,0.446668,2.965924,2.952682,4.992937,-0.543396,0.766725,0.376005,12.092588,10.758866
18190,morgajo02,1976,2B,113.0,27.0,111.0,60.0,0.320,151.0,472.0,0.444,266.0,599.0,321.0,2.273998,2.140820,1.423838,4.535192,0.866687,-0.000860,2.344281,3.225045,2.578104,4.113849,0.874159,-0.022024,11.241396,13.157461
24490,soriaal01,2002,2B,128.0,39.0,102.0,41.0,0.300,209.0,696.0,0.332,246.0,740.0,487.0,2.914982,1.788361,2.640873,2.852508,2.407766,2.062740,2.991519,2.778536,4.223291,2.524065,2.414939,1.993764,10.541750,12.938586
409,alomaro01,1999,2B,138.0,24.0,120.0,37.0,0.323,182.0,563.0,0.422,288.0,682.0,381.0,3.342304,2.493279,1.119580,2.498259,1.690367,0.745020,3.423011,3.671555,2.166807,2.189374,1.697679,0.706574,10.398770,12.441851
2766,boonebr01,2001,2B,118.0,37.0,141.0,5.0,0.331,206.0,623.0,0.372,255.0,685.0,417.0,2.487660,3.315684,2.438034,-0.335735,2.328055,1.192547,2.560027,4.713411,3.949093,-0.488157,2.335244,1.143733,9.041149,11.925884
18187,morgajo02,1973,2B,116.0,26.0,82.0,67.0,0.290,167.0,576.0,0.406,282.0,695.0,409.0,2.402195,1.005118,1.322419,5.155129,1.291812,1.093097,2.473728,1.786292,2.441005,4.699559,1.299202,1.046587,10.083576,11.653199


In [52]:
pitcher_years.groupby('POS').head(5).sort_values(['POS','Total_POS_Z'],ascending=[True,False])

,playerID,yearID,POS,W,SV,K,ERA,ERA_num,ERA_denom,WHIP,WHIP_num,WHIP_denom,W_LOG,SV_LOG,W_Z,SV_Z,K_Z,ERA_Z,WHIP_Z,Outs_Z,W_POS_Z,SV_POS_Z,K_POS_Z,ERA_POS_Z,WHIP_POS_Z,Outs_POS_Z,Total_Z,Total_POS_Z
8550,gagneer01,2003,RP,2.0,55.0,137.0,1.202,99.0,82.333333,0.717,59.0,82.333333,0.693147,4.007333,-0.530433,5.904298,0.855389,-2.547525,-3.409321,-0.969436,-0.155297,3.707347,2.946631,-2.153903,-3.018784,-0.272015,12.021457,11.810942
7055,eckerde01,1990,RP,4.0,48.0,73.0,0.614,45.0,73.333333,0.627,46.0,73.333333,1.386294,3.871201,-0.203492,5.098814,-0.475722,-3.107275,-3.886979,-1.120603,0.054647,3.135831,0.264175,-2.697439,-3.447975,-0.569796,12.101253,11.362116
13327,kimbrcr01,2012,RP,3.0,42.0,116.0,1.005,63.0,62.666667,0.654,41.0,62.666667,1.098612,3.737670,-0.339185,4.408399,0.418618,-2.735060,-3.743681,-1.299764,-0.032488,2.645960,2.066451,-2.336006,-3.319217,-0.922721,11.301886,10.924620
13332,kimbrcr01,2017,RP,5.0,35.0,126.0,1.435,99.0,69.000000,0.681,47.0,69.000000,1.609438,3.555348,-0.098240,3.602914,0.626604,-2.325719,-3.600384,-1.193387,0.122234,2.074444,2.485584,-1.938522,-3.190460,-0.713172,10.353068,10.156003
22059,rodnefe01,2012,RP,2.0,48.0,76.0,0.603,45.0,74.666667,0.790,59.0,74.666667,0.693147,3.871201,-0.530433,5.098814,-0.413326,-3.117747,-3.021887,-1.098208,-0.155297,3.135831,0.389915,-2.707607,-2.670662,-0.525680,10.735280,10.128098
13713,koufasa01,1965,SP,26.0,2.0,382.0,2.038,684.0,335.666667,0.867,291.0,335.666667,3.258097,0.693147,0.679393,-0.194370,5.951050,-1.751690,-2.613224,3.285627,1.587776,1.653995,5.340909,-2.084863,-2.847601,3.137281,12.916070,14.012746
8984,gibsobo01,1968,SP,22.0,0.0,268.0,1.123,342.0,304.666667,0.873,266.0,304.666667,3.091042,-25.328436,0.600598,-0.424509,3.580008,-2.622730,-2.581381,2.764942,1.344034,-0.282373,3.080286,-3.020334,-2.813276,2.555355,11.850481,12.352823
13711,koufasa01,1963,SP,25.0,0.0,306.0,1.881,585.0,311.000000,0.897,279.0,311.000000,3.218876,-25.328436,0.660894,-0.424509,4.370355,-1.901147,-2.454005,2.871318,1.530551,-0.282373,3.833827,-2.245375,-2.675978,2.674243,11.331854,11.825561
16594,mclaide01,1968,SP,31.0,0.0,280.0,1.955,657.0,336.000000,0.911,306.0,336.000000,3.433987,-25.328436,0.762357,-0.424509,3.829591,-1.830702,-2.379703,3.291226,1.844412,-0.282373,3.318246,-2.169720,-2.595887,3.143539,11.115632,11.702079
13714,koufasa01,1966,SP,27.0,0.0,317.0,1.728,558.0,323.000000,0.997,322.0,323.000000,3.295837,-25.328436,0.697194,-0.424509,4.599140,-2.046796,-1.923274,3.072874,1.642842,-0.282373,4.051957,-2.401798,-2.103901,2.899505,11.119388,11.622859


I mean it's not a crazy surprise, but wow A-Rod has the top 5 SS seasons of all time _and_ the top 2 3B seasons. And it's not even that he was dual-eligible in a season, that represents 7 different years. What a monster.

Ok, let's finally assemble our team! We want a typical 5x5 roster - C, 1B, 2B, SS, 3B, OF, OF, OF, Util. But remember we are placing a constraint on our team that no single human (ahem...A Rod) can be drafted twice.

In [53]:
non_OFs = position_years[
    position_years['POS'] != 'OF'
].groupby('POS').head(1)

In [54]:
three_OFs = position_years[
    position_years['POS'] == 'OF'
].head(3)

In [55]:
two_RPs = pitcher_years[
    pitcher_years['POS'] == 'RP'
].head(2)

two_SPs = pitcher_years[
    pitcher_years['POS'] == 'SP'
].head(2)

In [56]:
best_possible_draft = pd.concat([non_OFs, three_OFs, two_RPs, two_SPs]).sort_values('Total_POS_Z',ascending=False)

In [57]:
best_possible_draft

,playerID,yearID,POS,R,HR,RBI,SB,AVG,AVG_num,AVG_denom,OBP,OBP_num,OBP_denom,Outs,R_Z,RBI_Z,HR_Z,SB_Z,Hits_Z,Outs_Z,R_POS_Z,RBI_POS_Z,HR_POS_Z,SB_POS_Z,Hits_POS_Z,Outs_POS_Z,Total_Z,Total_POS_Z,W,SV,K,ERA,ERA_num,ERA_denom,WHIP,WHIP_num,WHIP_denom,W_LOG,SV_LOG,W_Z,SV_Z,K_Z,ERA_Z,WHIP_Z,W_POS_Z,SV_POS_Z,K_POS_Z,ERA_POS_Z,WHIP_POS_Z
22428,rodriiv01,1999,C,116.0,35.0,113.0,25.0,0.332,199.0,600.0,0.356,224.0,630.0,401.0,2.402195,2.219144,2.235195,1.435511,2.142062,0.993646,4.156318,3.109037,3.205023,6.552570,3.273639,1.889261,9.440462,18.407325,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
22369,rodrial01,2001,SS,133.0,52.0,135.0,18.0,0.318,201.0,632.0,0.399,292.0,732.0,431.0,3.128643,3.080711,3.959327,0.815575,2.195203,1.366586,3.297363,4.236547,5.734693,0.626347,2.184749,1.168135,11.812874,14.911563,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
22375,rodrial01,2007,3B,143.0,54.0,156.0,24.0,0.314,183.0,583.0,0.422,299.0,708.0,400.0,3.555966,3.903116,4.162166,1.346949,1.716937,0.981215,3.669584,3.823278,4.153101,2.432645,1.731781,0.920936,13.703919,14.889453,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
13713,koufasa01,1965,SP,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.285627,NaN,NaN,NaN,NaN,NaN,3.137281,12.916070,14.012746,26.0,2.0,382.0,2.038,684.0,335.666667,0.867,291.0,335.666667,3.258097,0.693147,0.679393,-0.194370,5.951050,-1.751690,-2.613224,1.587776,1.653995,5.340909,-2.084863,-2.847601
18190,morgajo02,1976,2B,113.0,27.0,111.0,60.0,0.320,151.0,472.0,0.444,266.0,599.0,321.0,2.273998,2.140820,1.423838,4.535192,0.866687,-0.000860,2.344281,3.225045,2.578104,4.113849,0.874159,-0.022024,11.241396,13.157461,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1136,bagweje01,1999,1B,143.0,42.0,126.0,30.0,0.304,171.0,562.0,0.454,331.0,729.0,391.0,3.555966,2.728252,2.945132,1.878323,1.398094,0.869333,3.537521,2.153163,2.303949,4.511768,1.280065,0.826303,11.636433,12.960163,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
27164,walkela01,1997,OF,143.0,49.0,130.0,33.0,0.366,208.0,568.0,0.452,300.0,664.0,360.0,3.555966,2.884900,3.655069,2.144010,2.381195,0.483962,3.345387,2.686922,3.391749,1.570544,2.268135,0.474226,14.137178,12.788511,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
24524,sosasa01,1998,OF,134.0,66.0,158.0,18.0,0.308,198.0,643.0,0.377,272.0,722.0,445.0,3.171375,3.981440,5.379201,0.815575,2.115492,1.540624,2.963441,3.740435,5.058869,0.442883,2.007783,1.507237,13.922459,12.706174,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2639,bondsba01,2001,OF,129.0,73.0,137.0,13.0,0.328,156.0,476.0,0.515,342.0,664.0,320.0,2.957714,3.159035,6.089138,0.372763,0.999539,-0.013291,2.751248,2.950300,5.745331,0.066996,0.914302,-0.011897,13.591481,12.440075,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8984,gibsobo01,1968,SP,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.764942,NaN,NaN,NaN,NaN,NaN,2.555355,11.850481,12.352823,22.0,0.0,268.0,1.123,342.0,304.666667,0.873,266.0,304.666667,3.091042,-25.328436,0.600598,-0.424509,3.580008,-2.622730,-2.581381,1.344034,-0.282373,3.080286,-3.020334,-2.813276


We need to pick an A Rod season to keep...I'm going to keep his absurd 2007 season at 3rd.

In [58]:
best_possible_draft.drop(22369,inplace=True)

In [59]:
non_a_rod_SS = position_years[
    (position_years['POS'] == 'SS') &
    (position_years['playerID'] != 'rodrial01')
].head(1)

In [60]:
best_possible_draft = pd.concat([best_possible_draft, non_a_rod_SS])

And we need any remaining player as our utility man - here we'll go back to our overall Z score and pick whoever is left that isn't already rostered.

In [61]:
position_years.sort_values('Total_Z', ascending=False, inplace=True)

In [62]:
util = position_years[
    ~position_years['playerID'].isin(best_possible_draft['playerID'])
].head(1)

In [63]:
three_ps = pitcher_years[
    ~pitcher_years['playerID'].isin(best_possible_draft['playerID'])
].head(3)

In [64]:
best_possible_draft = pd.concat([best_possible_draft, util, three_ps]).sort_values('POS',ascending=False)

In [65]:
best_possible_draft[['playerID','yearID','POS','Total_Z','Total_POS_Z']]

,playerID,yearID,POS,Total_Z,Total_POS_Z
21357,ramirha01,2007,SS,10.288166,12.173327
13713,koufasa01,1965,SP,12.916070,14.012746
8984,gibsobo01,1968,SP,11.850481,12.352823
16594,mclaide01,1968,SP,11.115632,11.702079
15891,martipe02,2000,SP,11.171983,11.547787
2293,bluevi01,1971,SP,10.859668,11.325996
8550,gagneer01,2003,RP,12.021457,11.810942
7055,eckerde01,1990,RP,12.101253,11.362116
27164,walkela01,1997,OF,14.137178,12.788511
24524,sosasa01,1998,OF,13.922459,12.706174


In [66]:
R, HR, RBI, SB, AVG_num, AVG_denom = best_possible_draft[['R','HR','RBI','SB','AVG_num','AVG_denom']].sum().values

In [67]:
R = int(R)
HR = int(HR)
RBI = int(RBI)
SB = int(SB)
AVG = round(AVG_num / AVG_denom,3)
print(R,HR,RBI,SB,AVG)

1188 415 1140 286 0.328


In [68]:
W, SV, K, ER, WH, I = best_possible_draft[['W','SV','K','ERA_num','WHIP_num','ERA_denom']].sum().values

In [69]:
W = int(W)
SV = int(SV)
K = int(K)
ERA = round(ER / I,3)
WHIP = round(WH / I,3)
print(W, SV, K, ERA, WHIP)

127 105 1725 1.669 0.86


So there we have it. Our truly absurd line-up:

```
* C: Ivan Rodriguez  '99 : 116 R, 113 RBI, 35 HR, 25 SB, .332 AVG
* 1B: Mark McGwire   '98 : 130 R, 147 RBI, 70 HR, 1  SB, .299 AVG
* 2B: Roberto Alomar '99 : 138 R, 120 RBI, 24 HR, 37 SB, .323 AVG
* 3B: Alex Rodriguez '07 : 143 R, 156 RBI, 54 HR, 23 SB, .314 AVG
* SS: Hanley Ramirez '07 : 125 R, 81  RBI, 29 HR, 51 SB, .332 AVG
* OF: Larry Walker   '97 : 143 R, 130 RBI, 49 HR, 33 SB, .366 AVG
* OF: Sammy Sosa     '98 : 134 R, 158 RBI, 66 HR, 18 SB, .308 AVG
* OF: Barry Bonds    '01 : 129 R, 137 RBI, 73 HR, 13 SB, .328 AVG
* Util: Ellis Burks  '96 : 142 R, 128 RBI, 40 HR, 32 SB, .344 AVG
```

All together they'd score 1200 runs, knock in 1170 RBIs, hit 440 homers, steal 234 bases while hitting a ridiculous .327 average.

In [70]:
all_players = pd.concat([position_years, pitcher_years]).sort_values('POS',ascending=False)

In [71]:
all_players = all_players.fillna(0)

#### Try some drafts - first, let's just fully automate one

In [72]:
class Draft():
    def __init__(self, player_pool = None, db = None, teams = 12, autodraft_level = 50):
        assert db is not None
        assert player_pool is not None
        assert autodraft_level > 0 and autodraft_level < 101
        
        self.db = db
        self.player_pool = player_pool.sort_values('Total_POS_Z',ascending=False)
        
        self.drafted = set()
        self.draftees = []
        
        if type(teams) == int:
            self.teams = [Team(name = f'Team {i}', draft = self) for i in range(teams)]
        elif type(teams) == list:
            self.teams = [Team(name = name, draft = self) for name in teams]
            
        self.active_team_ix = 0
        self.snake_up = True
        self.pick_ix = 1
        self.complete = False
        
        self.autodraft_depth = 301 - (3 * autodraft_level)
        
        
    def bump_active_team(self):
        if any([
            any([
                len(team.roster[pos]['picks']) < team.roster[pos]['slots']
                for pos in team.roster
            ]) for team in self.teams
        ]):
            if self.snake_up:
                self.active_team_ix += 1
                if self.active_team_ix == len(self.teams):
                    self.snake_up = False
                    self.active_team_ix += -1

            else:
                self.active_team_ix += -1
                if self.active_team_ix == -1:
                    self.snake_up = True
                    self.active_team_ix += 1

            self.pick_ix += 1
        else:
            self.complete = True
            
        
    def make_auto_pick(self):
        self.teams[self.active_team_ix].pick(auto=True)
        self.bump_active_team()
                
    def run_full_autodraft(self):
        while self.complete == False:
            self.make_auto_pick()
        return self
            
    def run_until_human_pick(self):
        while self.teams[self.active_team_ix].auto == True and self.complete == False:
            self.make_auto_pick()
        return self
    
    def standings(self):
        team_totals = [team.totals() for team in self.teams]
        standings = pd.DataFrame(team_totals, columns=['Name','R','HR','RBI','SB','AVG','W','SV','K','ERA','WHIP'])
        standings.set_index('Name', inplace=True)
        ranks = standings.rank()
        ranks.rename(columns = {
            'R':'R_p',
            'HR':'HR_p',
            'RBI':'RBI_p',
            'SB':'SB_p',
            'AVG':'AVG_p',
            'W':'W_p',
            'SV':'SV_p',
            'K':'K_p',
            'ERA':'ERA_p',
            'WHIP':'WHIP_p'
        }, inplace = True)
        ranks['ERA_p'] = (1 + len(self.teams)) - ranks['ERA_p']
        ranks['WHIP_p'] = (1 + len(self.teams)) - ranks['WHIP_p']
        ranks['Total'] = ranks.sum(axis=1)
        ranks = pd.merge(
            ranks, standings, 
            how='left',
            left_index=True, right_index=True
        ).sort_values('Total',ascending=False)
        ranks = ranks[[
            'R', 'R_p',
            'HR', 'HR_p',
            'RBI', 'RBI_p',
            'SB', 'SB_p',
            'AVG', 'AVG_p',
            'W', 'W_p',
            'SV', 'SV_p',
            'K', 'K_p',
            'ERA', 'ERA_p',
            'WHIP', 'WHIP_p',
            'Total'
        ]]
        return ranks
    
    def __repr__(self):
        return f'{chr(10).join(self.draftees)} \n\n {self.teams[self.active_team_ix]}'
    
    def __str__(self):
        return f'{chr(10).join(self.draftees)} \n\n {self.teams[self.active_team_ix]}'

In [73]:
class Team():
    def __init__(self, name = None, draft = None):
        assert draft is not None
        self.roster = {
            'C'    : {'picks': [], 'slots': 1},
            '1B'   : {'picks': [], 'slots': 1},
            '2B'   : {'picks': [], 'slots': 1},
            'SS'   : {'picks': [], 'slots': 1},
            '3B'   : {'picks': [], 'slots': 1},
            'OF'   : {'picks': [], 'slots': 3},
            'Util' : {'picks': [], 'slots': 1},
            'SP'   : {'picks': [], 'slots': 2},
            'RP'   : {'picks': [], 'slots': 2},
            'P'    : {'picks': [], 'slots': 3},
        }
        self.draft = draft
        self.name = name
        self.auto = True
        
    def suggest_pick(self):
        empty_slots = [
            pos for pos in self.roster if ((len(self.roster[pos]['picks']) < self.roster[pos]['slots']) and pos not in ['Util','P'])
        ]

        if len(empty_slots) == 0:
            if self.roster['Util']['picks'] == []:
                candidates = self.draft.player_pool[
                    (~self.draft.player_pool['playerID'].isin(self.draft.drafted)) &
                    (~self.draft.player_pool['POS'].isin(['SP','RP']))
                ].sort_values('Total_Z', ascending=False)
            else:
                candidates = self.draft.player_pool[
                    (~self.draft.player_pool['playerID'].isin(self.draft.drafted)) &
                    (self.draft.player_pool['POS'].isin(['SP','RP']))
                ].sort_values('Total_Z', ascending=False)
        else:
            candidates = self.draft.player_pool[
                (self.draft.player_pool['POS'].isin(empty_slots)) &
                (~self.draft.player_pool['playerID'].isin(self.draft.drafted))
            ].sort_values('Total_POS_Z', ascending=False)

        pick = candidates.head(self.draft.autodraft_depth).sample(1)
        
        playerID = pick['playerID'].item()
        yearID = pick['yearID'].item()
        POS = pick['POS'].item()
        
        if len(self.roster[POS]['picks']) == self.roster[POS]['slots']:
            if POS in ('SP','RP'):
                POS = 'P'
            else:
                POS = 'Util'
        elif len(self.roster[POS]['picks']) > self.roster[POS]['slots']:
            assert False, ("Should not be here!", POS, self.roster[POS])

        return (playerID, yearID, POS)
        
    def pick(self, playerID = None, name = None, yearID = None, POS = None, auto = False):
        if self.auto or auto:
            (playerID, yearID, POS) = self.suggest_pick()
            
        if playerID == None or playerID == '':
            assert (name != None and name != '')
            name_lookup = name_to_playerID(name)
            if len(name_lookup) > 1:
                possible_picks = []
                for (name, playerID) in name_lookup:
                    draftables = self.draft.player_pool[(self.draft.player_pool['playerID'] == playerID)]
                    if len(draftables) == 0:
                        possible_picks.append((name, playerID, 'No eligible seasons'))
                        
                    else:
                        POS = draftables['POS'].unique()
                        min_year = min(draftables['yearID'].unique())
                        max_year = max(draftables['yearID'].unique())
                        possible_picks.append((name, playerID, POS, f'{min_year}-{max_year}'))
                return f'More than 1 player with that name, possible playerIDs are: {possible_picks}'
            elif len(name_lookup) == 0:
                return f'No players found with that name.'
            else:
                playerID = name_lookup[0][1]
                
        if playerID in self.draft.drafted:
            print(self.draft.draftees)
            return f'Player {playerID_to_name(playerID, self.draft.db)} already drafted.'
        
        if len(self.draft.player_pool[
            (self.draft.player_pool['playerID'] == playerID) &
            (self.draft.player_pool['yearID'] == yearID) &
            (self.draft.player_pool['POS'] == POS)
        ]) < 1 and POS != 'Util' and POS != 'P':
            return f'Player {playerID_to_name(playerID, self.draft.db)} not eligible for {POS} in year {yearID}.'
        
        if len(self.roster[POS]['picks']) >= self.roster[POS]['slots']:
            return f'Too many players already drafted at {POS}.'
            
        draftee = Draftee(playerID, yearID, POS, self.draft)
        self.roster[draftee.POS]['picks'].append(draftee)
        self.draft.drafted.add(playerID)
        self.draft.draftees.extend([f'#{self.draft.pick_ix:>2} - {self.name:6}: {draftee}'])
        if auto == False and self.auto == False:
            self.draft.bump_active_team()
            self.draft.run_until_human_pick()
        return self.draft
    
    def totals(self):
        player_pos = [self.roster[pos]['picks'] for pos in self.roster]
        players = [player for pos in player_pos for player in pos]
        R = int(sum([player.stats['R'] for player in players]))
        RBI = int(sum([player.stats['RBI'] for player in players]))
        HR = int(sum([player.stats['HR'] for player in players]))
        SB = int(sum([player.stats['SB'] for player in players]))
        AVG_num = sum([player.stats['AVG_num'] for player in players])
        AVG_denom = sum([player.stats['AVG_denom'] for player in players])
        if AVG_denom > 0:
            AVG = round(AVG_num / AVG_denom, 3)
        else:
            AVG = 0.000
            
        W = int(sum([player.stats['W'] for player in players]))
        SV = int(sum([player.stats['SV'] for player in players]))
        K = int(sum([player.stats['K'] for player in players]))
        ERA_num = int(sum([player.stats['ERA_num'] for player in players]))
        WHIP_num = int(sum([player.stats['WHIP_num'] for player in players]))
        ERA_denom = int(sum([player.stats['ERA_denom'] for player in players]))
        if ERA_denom > 0:
            ERA = round(ERA_num / ERA_denom, 2)
            WHIP = round(WHIP_num / ERA_denom, 2)
        else:
            ERA = 0.000
            WHIP = 0.000
            
        return (self.name,R,HR,RBI,SB,AVG,W,SV,K,ERA,WHIP)
        
    def __repr__(self):
        roster_dict = {pos: str(self.roster[pos]['picks']) for pos in self.roster}
        return f'{self.name}:\n\n{json.dumps(roster_dict, sort_keys=True, indent=2)}'

In [74]:
class Draftee():
    def __init__(self, playerID, yearID, POS, draft):
        self.playerID = playerID
        self.name = playerID_to_name(playerID, draft.db)
        self.yearID = yearID
        self.POS = POS
        self.draft = draft
        self.stats = draft.player_pool[
            (draft.player_pool['playerID'] == playerID) &
            (draft.player_pool['yearID'] == yearID)
        ].iloc[0].to_dict()
        
    def __repr__(self):
        if self.POS in ['SP','RP','P']:
            return f'{self.name:18} {self.yearID} {self.POS:4}: {int(self.stats["W"]):>2} W, {int(self.stats["SV"]):>2} SV, {int(self.stats["K"]):>3} K, {self.stats["ERA"]:>.2f} ERA, {self.stats["WHIP"]:>.3f} WHIP, {self.stats["Total_Z"]:>.2f} Z, {self.stats["Total_POS_Z"]:>.2f} POS_Z'
        else:
            return f'{self.name:18} {self.yearID} {self.POS:4}: {int(self.stats["R"]):>3} R, {int(self.stats["RBI"]):>3} RBI, {int(self.stats["HR"]):>2} HR, {int(self.stats["SB"]):>3} SB, {self.stats["AVG"]:.3f} AVG, {self.stats["Total_Z"]:>.2f} Z, {self.stats["Total_POS_Z"]:>.2f} POS_Z'
    
    def __str__(self):
        if self.POS in ['SP','RP','P']:
            return f'{self.name:18} {self.yearID} {self.POS:4}: {int(self.stats["W"]):>2} W, {int(self.stats["SV"]):>2} SV, {int(self.stats["K"]):>3} K, {self.stats["ERA"]:>.2f} ERA, {self.stats["WHIP"]:>.3f} WHIP, {self.stats["Total_Z"]:>.2f} Z, {self.stats["Total_POS_Z"]:>.2f} POS_Z'
        else:
            return f'{self.name:18} {self.yearID} {self.POS:4}: {int(self.stats["R"]):>3} R, {int(self.stats["RBI"]):>3} RBI, {int(self.stats["HR"]):>2} HR, {int(self.stats["SB"]):>3} SB, {self.stats["AVG"]:.3f} AVG, {self.stats["Total_Z"]:>.2f} Z, {self.stats["Total_POS_Z"]:>.2f} POS_Z'
   

In [75]:
#TODO

# class Draftable...DB schema for this that maps to what I've been calling "player_pool"?

In [76]:
example = Draft(player_pool = all_players, db = db, teams = 6, autodraft_level = 85)

In [77]:
example.run_full_autodraft()

# 1 - Team 0: Alfonso Soriano    2002 2B  : 128 R, 102 RBI, 39 HR,  41 SB, 0.300 AVG, 10.54 Z, 12.94 POS_Z
# 2 - Team 1: Andres Galarraga   1996 1B  : 119 R, 150 RBI, 47 HR,  18 SB, 0.304 AVG, 10.94 Z, 11.12 POS_Z
# 3 - Team 2: Roberto Alomar     1999 2B  : 138 R, 120 RBI, 24 HR,  37 SB, 0.323 AVG, 10.40 Z, 12.44 POS_Z
# 4 - Team 3: Jose Ramirez       2018 2B  : 110 R, 105 RBI, 39 HR,  34 SB, 0.270 AVG, 8.67 Z, 11.11 POS_Z
# 5 - Team 4: Alex Rodriguez     1998 SS  : 123 R, 124 RBI, 42 HR,  46 SB, 0.310 AVG, 12.22 Z, 14.87 POS_Z
# 6 - Team 5: Jason Kendall      2000 C   : 112 R,  58 RBI, 14 HR,  22 SB, 0.320 AVG, 4.44 Z, 11.39 POS_Z
# 7 - Team 5: Howard Johnson     1989 SS  : 104 R, 101 RBI, 36 HR,  41 SB, 0.287 AVG, 8.97 Z, 11.24 POS_Z
# 8 - Team 4: Jeff Bagwell       1999 1B  : 143 R, 126 RBI, 42 HR,  30 SB, 0.304 AVG, 11.64 Z, 12.96 POS_Z
# 9 - Team 3: Sandy Koufax       1965 SP  : 26 W,  2 SV, 382 K, 2.04 ERA, 0.867 WHIP, 12.92 Z, 14.01 POS_Z
#10 - Team 2: Luis Tiant         1968 SP

In [78]:
example.teams

[Team 0:
 
 {
   "1B": "[Albert Pujols      2009 1B  : 124 R, 135 RBI, 47 HR,  16 SB, 0.327 AVG, 10.95 Z, 11.03 POS_Z]",
   "2B": "[Alfonso Soriano    2002 2B  : 128 R, 102 RBI, 39 HR,  41 SB, 0.300 AVG, 10.54 Z, 12.94 POS_Z]",
   "3B": "[Adrian Beltre      2004 3B  : 104 R, 121 RBI, 48 HR,   7 SB, 0.334 AVG, 9.03 Z, 9.39 POS_Z]",
   "C": "[Russell Martin     2007 C   :  87 R,  87 RBI, 19 HR,  21 SB, 0.293 AVG, 4.35 Z, 11.08 POS_Z]",
   "OF": "[Albert Belle       1995 OF  : 121 R, 126 RBI, 50 HR,   5 SB, 0.317 AVG, 9.57 Z, 8.63 POS_Z, Jacoby Ellsbury    2011 OF  : 119 R, 105 RBI, 32 HR,  39 SB, 0.321 AVG, 9.95 Z, 8.65 POS_Z, Orlando Cepeda     1961 OF  : 105 R, 142 RBI, 46 HR,  12 SB, 0.311 AVG, 9.59 Z, 9.28 POS_Z]",
   "P": "[Johan Santana      2004 P   : 20 W,  0 SV, 265 K, 2.60 ERA, 0.921 WHIP, 8.54 Z, 8.87 POS_Z, Sam McDowell       1968 P   : 15 W,  0 SV, 283 K, 1.81 ERA, 1.115 WHIP, 8.81 Z, 9.05 POS_Z, Mickey Lolich      1971 P   : 25 W,  0 SV, 308 K, 2.92 ERA, 1.144 WHIP, 8.91 Z,

In [79]:
example.standings()

,R,R_p,HR,HR_p,RBI,RBI_p,SB,SB_p,AVG,AVG_p,W,W_p,SV,SV_p,K,K_p,ERA,ERA_p,WHIP,WHIP_p,Total
Name,,,,,,,,,,,,,,,,,,,,,
Team 2,1106,6.0,358,5.0,1101,6.0,155,2.0,0.324,6.0,126,6.0,52,1.0,1679,6.0,2.00,3.0,0.97,2.0,43.0
Team 3,1053,2.0,332,3.5,1010,4.0,219,3.0,0.312,3.0,114,5.0,84,2.0,1648,5.0,1.80,5.0,0.96,3.0,35.5
Team 5,1100,5.0,395,6.0,1076,5.0,140,1.0,0.311,2.0,110,3.5,95,4.0,1641,4.0,2.14,1.0,0.93,4.0,35.5
Team 4,1061,3.0,300,1.0,1005,3.0,362,6.0,0.321,5.0,75,1.0,174,5.0,1207,2.0,1.89,4.0,0.92,5.0,35.0
Team 1,1062,4.0,312,2.0,987,1.0,231,4.0,0.308,1.0,78,2.0,176,6.0,933,1.0,1.65,6.0,0.89,6.0,33.0
Team 0,1045,1.0,332,3.5,994,2.0,260,5.0,0.314,4.0,110,3.5,92,3.0,1423,3.0,2.11,2.0,1.00,1.0,28.0


In [80]:
example.complete

True

#### Do a manual draft?

In [81]:
two_man_draft = Draft(position_years, db, teams=['Nick','Dad'])

In [82]:
nick = two_man_draft.teams[0]
dad = two_man_draft.teams[1]

In [83]:
draft_widget = widgets.interact.options(manual=True, manual_name="Draft!")

In [84]:
nick.roster['Util'] = []

In [85]:
draft_widget(
    nick.pick,
    playerID = '',
    name = '',
    yearID = widgets.IntSlider(min=1985, max=2015, step=1, value=2007),
    POS = ['C','1B','2B','SS','3B','OF','Util'],
    auto=widgets.fixed(False)
)

interactive(children=(Text(value='', description='playerID'), Text(value='', description='name'), IntSlider(va…

<function ipywidgets.widgets.interaction._InteractFactory.__call__.<locals>.<lambda>(*args, **kwargs)>

In [86]:
draft_widget(
    dad.pick,
    playerID = '',
    name = '',
    yearID = widgets.IntSlider(min=1985, max=2015, step=1, value=2007),
    POS = ['C','1B','2B','SS','3B','OF','Util'],
    auto=widgets.fixed(False)
)

interactive(children=(Text(value='', description='playerID'), Text(value='', description='name'), IntSlider(va…

<function ipywidgets.widgets.interaction._InteractFactory.__call__.<locals>.<lambda>(*args, **kwargs)>

In [87]:
name_to_playerID('Baerga')

[('Carlos Baerga', 'baergca01')]

In [88]:
position_years[position_years['playerID'] == 'gonzalu01'].sort_values('yearID')

,playerID,yearID,POS,R,HR,RBI,SB,AVG,AVG_num,AVG_denom,OBP,OBP_num,OBP_denom,Outs,R_Z,RBI_Z,HR_Z,SB_Z,Hits_Z,Outs_Z,R_POS_Z,RBI_POS_Z,HR_POS_Z,SB_POS_Z,Hits_POS_Z,Outs_POS_Z,Total_Z,Total_POS_Z
9813,gonzalu01,1991,OF,51.0,13.0,69.0,10.0,0.254,120.0,473.0,0.320,168.0,525.0,353.0,-0.375400,0.496010,0.003965,0.107076,0.043008,0.396943,-0.558956,0.391769,-0.138624,-0.158536,-0.022966,0.389154,-0.122284,-0.876468
9814,gonzalu01,1992,OF,40.0,10.0,55.0,7.0,0.243,94.0,387.0,0.289,120.0,415.0,293.0,-0.845455,-0.052259,-0.300294,-0.158610,-0.647820,-0.348937,-1.025780,-0.134988,-0.432822,-0.384068,-0.699883,-0.340030,-1.655503,-2.337510
9815,gonzalu01,1993,OF,82.0,15.0,72.0,20.0,0.300,162.0,540.0,0.361,219.0,607.0,378.0,0.949299,0.613497,0.206804,0.992700,1.158961,0.707726,0.756638,0.504645,0.057508,0.593238,1.070514,0.692981,3.213534,2.289561
9816,gonzalu01,1994,OF,57.0,8.0,67.0,15.0,0.273,107.0,392.0,0.353,159.0,450.0,285.0,-0.119007,0.417686,-0.503133,0.549888,-0.302406,-0.448387,-0.304325,0.316518,-0.628954,0.217351,-0.361424,-0.437254,0.491415,-0.323580
9817,gonzalu01,1995,OF,69.0,13.0,69.0,6.0,0.276,130.0,471.0,0.357,193.0,540.0,341.0,0.393780,0.496010,0.003965,-0.247173,0.308711,0.247767,0.204937,0.391769,-0.138624,-0.459245,0.237386,0.243318,0.707526,-0.007095
9818,gonzalu01,1996,OF,70.0,15.0,79.0,9.0,0.271,131.0,483.0,0.354,196.0,554.0,352.0,0.436512,0.887632,0.206804,0.018514,0.335281,0.384511,0.247376,0.768023,0.057508,-0.233713,0.263421,0.377001,1.500231,0.725614
9819,gonzalu01,1997,OF,78.0,10.0,68.0,10.0,0.258,142.0,550.0,0.345,218.0,631.0,408.0,0.778370,0.456848,-0.300294,0.107076,0.627555,1.080665,0.586884,0.354143,-0.432822,-0.158536,0.549809,1.057573,0.588890,-0.158095
9820,gonzalu01,1998,OF,84.0,23.0,71.0,12.0,0.267,146.0,547.0,0.340,211.0,620.0,401.0,1.034763,0.574335,1.018160,0.284201,0.733836,0.993646,0.841515,0.467020,0.842035,-0.008181,0.653950,0.972502,2.651649,1.823837
9821,gonzalu01,1999,OF,112.0,26.0,111.0,9.0,0.336,206.0,614.0,0.403,279.0,692.0,408.0,2.231266,2.140820,1.322419,0.018514,2.328055,1.080665,2.029793,1.972038,1.136233,-0.233713,2.216064,1.057573,6.960408,6.062842
9822,gonzalu01,2000,OF,106.0,31.0,114.0,2.0,0.311,192.0,618.0,0.392,282.0,720.0,426.0,1.974873,2.258306,1.829517,-0.601422,1.956070,1.304429,1.775162,2.084915,1.626562,-0.759955,1.851571,1.276329,6.112915,5.301927


In [89]:
two_man_draft.standings()

TypeError: list indices must be integers or slices, not str

In [ ]:
nick

In [ ]:
dad

#### Draft against the computer?

In [81]:
against_computer = Draft(player_pool = all_players, db = db, teams = 12, autodraft_level = 60)
my_team = against_computer.teams[random.randint(0,11)]
my_team.auto = False
my_team.name = 'Nick'

In [82]:
against_computer.run_until_human_pick()

# 1 - Team 0: Alex Rodriguez     2003 SS  : 124 R, 118 RBI, 47 HR,  17 SB, 0.298 AVG, 9.70 Z, 12.51 POS_Z
# 2 - Team 1: Ken Griffey        1997 OF  : 125 R, 147 RBI, 56 HR,  15 SB, 0.304 AVG, 11.76 Z, 10.63 POS_Z
# 3 - Team 2: Barry Bonds        1993 OF  : 129 R, 123 RBI, 46 HR,  29 SB, 0.336 AVG, 11.91 Z, 10.66 POS_Z
# 4 - Team 3: Francisco Lindor   2018 SS  : 129 R,  92 RBI, 38 HR,  25 SB, 0.277 AVG, 8.10 Z, 10.43 POS_Z
# 5 - Team 4: Alfonso Soriano    2002 2B  : 128 R, 102 RBI, 39 HR,  41 SB, 0.300 AVG, 10.54 Z, 12.94 POS_Z 

 Nick:

{
  "1B": "[]",
  "2B": "[]",
  "3B": "[]",
  "C": "[]",
  "OF": "[]",
  "P": "[]",
  "RP": "[]",
  "SP": "[]",
  "SS": "[]",
  "Util": "[]"
}

In [83]:
draft_widget = widgets.interact.options(manual=True, manual_name="Draft!")

In [84]:
draft_widget(
    my_team.pick,
    playerID = '',
    name = '',
    yearID = widgets.IntSlider(min=1961, max=2019, step=1, value=2007),
    POS = ['C','1B','2B','SS','3B','OF','Util','SP','RP','P'],
    auto=widgets.fixed(False)
)

interactive(children=(Text(value='', description='playerID'), Text(value='', description='name'), IntSlider(va…

<function ipywidgets.widgets.interaction._InteractFactory.__call__.<locals>.<lambda>(*args, **kwargs)>

In [89]:
against_computer.standings()

,R,R_p,HR,HR_p,RBI,RBI_p,SB,SB_p,AVG,AVG_p,W,W_p,SV,SV_p,K,K_p,ERA,ERA_p,WHIP,WHIP_p,Total
Name,,,,,,,,,,,,,,,,,,,,,
Nick,1109,12.0,344,9.0,1012,4.0,256,10.0,0.323,12.0,115,10.0,97,7.0,1303,7.0,2.11,9.0,0.96,10.0,90.0
Team 10,989,2.0,315,5.0,1131,12.0,120,2.0,0.318,9.5,121,12.0,76,2.0,1591,12.0,1.98,11.0,0.97,9.0,76.5
Team 1,1034,7.5,334,7.0,1081,11.0,157,6.0,0.311,6.0,79,2.0,115,10.0,1134,2.0,1.54,12.0,0.90,11.5,75.0
Team 4,1086,11.0,323,6.0,1037,6.0,212,8.0,0.318,9.5,69,1.0,151,12.0,1094,1.0,2.23,5.5,0.98,8.0,68.0
Team 8,1034,7.5,343,8.0,1051,7.0,149,4.0,0.322,11.0,100,5.5,106,8.5,1172,4.0,2.29,4.0,1.01,5.5,65.0
Team 0,1013,6.0,373,12.0,1072,9.0,137,3.0,0.302,4.0,111,9.0,85,3.5,1493,11.0,2.36,1.5,1.01,5.5,64.5
Team 3,998,4.0,351,10.0,1073,10.0,114,1.0,0.316,8.0,100,5.5,127,11.0,1389,8.0,2.23,5.5,1.09,1.0,64.0
Team 7,1046,10.0,310,4.0,1054,8.0,151,5.0,0.315,7.0,118,11.0,69,1.0,1299,6.0,2.12,8.0,1.03,2.5,62.5
Team 6,995,3.0,295,2.0,937,2.0,286,11.0,0.308,5.0,88,4.0,85,3.5,1425,9.0,2.02,10.0,0.90,11.5,61.0


In [91]:
against_computer.player_pool[
    (~against_computer.player_pool['playerID'].isin(against_computer.drafted))
].sort_values('Total_POS_Z',ascending=False).head(25)

,playerID,yearID,POS,R,HR,RBI,SB,AVG,AVG_num,AVG_denom,OBP,OBP_num,OBP_denom,Outs,R_Z,RBI_Z,HR_Z,SB_Z,Hits_Z,Outs_Z,R_POS_Z,RBI_POS_Z,HR_POS_Z,SB_POS_Z,Hits_POS_Z,Outs_POS_Z,Total_Z,Total_POS_Z,W,SV,K,ERA,ERA_num,ERA_denom,WHIP,WHIP_num,WHIP_denom,W_LOG,SV_LOG,W_Z,SV_Z,K_Z,ERA_Z,WHIP_Z,W_POS_Z,SV_POS_Z,K_POS_Z,ERA_POS_Z,WHIP_POS_Z
8197,fiskca01,1985,C,85.0,37.0,107.0,17.0,0.238,129.0,543.0,0.320,198.0,618.0,414.0,1.077496,1.984171,2.438034,0.727013,0.282141,1.155253,2.367963,2.810931,3.476024,4.249639,0.969534,2.098247,5.353601,11.775843,0.0,0.0,0.0,0.000,0.0,0.000000,0.000,0.0,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
22561,rolliji01,2007,SS,139.0,30.0,94.0,41.0,0.296,212.0,716.0,0.344,268.0,778.0,504.0,3.385037,1.475064,1.728097,2.852508,2.487476,2.274072,3.558102,2.282223,2.873462,2.603585,2.474787,2.039525,9.654110,11.752634,0.0,0.0,0.0,0.000,0.0,0.000000,0.000,0.0,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
16462,martiru01,2007,C,87.0,19.0,87.0,21.0,0.293,158.0,540.0,0.374,232.0,620.0,382.0,1.162960,1.200929,0.612482,1.081262,1.052680,0.757451,2.483340,1.817245,1.037011,5.401104,1.924092,1.583819,4.352861,11.078973,0.0,0.0,0.0,0.000,0.0,0.000000,0.000,0.0,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
14026,kentje01,2000,2B,114.0,33.0,125.0,12.0,0.334,196.0,587.0,0.424,295.0,695.0,391.0,2.316731,2.689090,2.032356,0.284201,2.062351,0.869333,2.387430,3.919616,3.400697,0.097553,2.069592,0.828007,8.515396,11.046880,0.0,0.0,0.0,0.000,0.0,0.000000,0.000,0.0,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
9327,goodedw01,1985,SP,0.0,0.0,0.0,0.0,0.000,0.0,0.0,0.000,0.0,0.0,0.0,0.000000,0.000000,0.000000,0.000000,0.000000,2.294645,0.000000,0.000000,0.000000,0.000000,0.000000,2.029745,10.213876,10.675947,24.0,0.0,268.0,1.529,423.0,276.666667,0.980,271.0,276.666667,3.178054,-2.532844e+01,0.641639,-0.424509,3.580008,-2.236235,-2.013498,1.470989,-0.282373,3.080286,-2.605251,-2.201154
27378,wathajo01,1982,C,79.0,3.0,51.0,36.0,0.270,121.0,448.0,0.343,171.0,499.0,327.0,0.821102,-0.208908,-1.010231,2.409697,0.069578,0.073728,2.021829,0.028610,-1.131001,9.719101,0.706208,0.699646,2.007510,10.645100,0.0,0.0,0.0,0.000,0.0,0.000000,0.000,0.0,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
13113,jeterde01,1999,SS,134.0,24.0,102.0,19.0,0.349,219.0,627.0,0.438,322.0,736.0,408.0,3.171375,1.788361,1.119580,0.904137,2.673469,1.080665,3.340819,2.663554,2.093126,0.712314,2.659358,0.893588,8.576256,10.575583,0.0,0.0,0.0,0.000,0.0,0.000000,0.000,0.0,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
8189,fiskca01,1977,C,106.0,26.0,102.0,7.0,0.315,169.0,536.0,0.402,253.0,630.0,367.0,1.974873,1.788361,1.322419,-0.158610,1.344953,0.570981,3.579429,2.562510,1.985516,1.370975,2.286165,1.342681,5.701014,10.441913,0.0,0.0,0.0,0.000,0.0,0.000000,0.000,0.0,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
1824,bellja01,1999,2B,132.0,38.0,112.0,7.0,0.289,170.0,589.0,0.374,256.0,684.0,419.0,3.085911,2.179982,2.539453,-0.158610,1.371523,1.217410,3.164116,3.274658,4.086192,-0.320811,1.378897,1.168020,7.800849,10.415031,0.0,0.0,0.0,0.000,0.0,0.000000,0.000,0.0,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
15683,maricju01,1965,SP,0.0,0.0,0.0,0.0,0.000,0.0,0.0,0.000,0.0,0.0,0.0,0.000000,0.000000,0.000000,0.000000,0.000000,2.608176,0.000000,0.000000,0.000000,0.000000,0.000000,2.380152,9.658912,10.397181,22.0,1.0,240.0,2.133,630.0,295.333333,0.928,274.0,295.333333,3.091042,1.000000e-11,0.60

In [90]:
my_team

Nick:

{
  "1B": "[Frank Thomas       1996 1B  : 110 R, 134 RBI, 40 HR,   1 SB, 0.349 AVG, 8.71 Z, 7.55 POS_Z]",
  "2B": "[Chase Utley        2006 2B  : 131 R, 102 RBI, 32 HR,  15 SB, 0.309 AVG, 7.90 Z, 10.16 POS_Z]",
  "3B": "[Vinny Castilla     1998 3B  : 108 R, 144 RBI, 46 HR,   5 SB, 0.319 AVG, 9.37 Z, 9.65 POS_Z]",
  "C": "[Ivan Rodriguez     1999 C   : 116 R, 113 RBI, 35 HR,  25 SB, 0.332 AVG, 9.44 Z, 18.41 POS_Z]",
  "OF": "[Larry Walker       1997 OF  : 143 R, 130 RBI, 49 HR,  33 SB, 0.366 AVG, 14.14 Z, 12.79 POS_Z, Jose Canseco       1988 OF  : 120 R, 124 RBI, 42 HR,  40 SB, 0.307 AVG, 11.49 Z, 10.11 POS_Z, Mike Trout         2012 OF  : 129 R,  83 RBI, 30 HR,  49 SB, 0.326 AVG, 10.29 Z, 8.88 POS_Z]",
  "P": "[Roy Halladay       2010 P   : 21 W,  0 SV, 219 K, 2.44 ERA, 1.045 WHIP, 7.45 Z, 7.80 POS_Z, Dennis Eckersley   1990 P   :  4 W, 48 SV,  73 K, 0.61 ERA, 0.627 WHIP, 12.10 Z, 11.36 POS_Z, Denny McLain       1968 P   : 31 W,  0 SV, 280 K, 1.96 ERA, 0.911 WHIP, 11.12 Z, 11.70

In [122]:
my_team.roster['3B']['picks'].append(my_team.roster['Util']['picks'][0])

In [ ]:
standings = against_computer.standings()

In [ ]:
standings

In [ ]:
def draft_to_player_results(draft):
    player_results = []

    standings = draft.standings()
    
    for team in draft.teams:
        results = standings.loc[team.name]
        total_p = results['Total']
        for pos in team.roster:
            for player in team.roster[pos]['picks']:
                total_z = player.stats['Total_Z']
                total_pos_z = player.stats['Total_POS_Z']
                if player.POS in ('SP','RP','P'):
                    total_hmean = hmean([
                        stat for stat in
                        [
                            player.stats['W_Z'], player.stats['SV_Z'], player.stats['K_Z'], -1 * player.stats['ERA_Z'], -1 * player.stats['WHIP_Z']
                        ] if stat >= 0
                    ])
                    total_pos_hmean = hmean([
                        stat for stat in
                        [
                            player.stats['W_POS_Z'], player.stats['SV_POS_Z'], player.stats['K_POS_Z'], -1 * player.stats['ERA_POS_Z'], -1 * player.stats['WHIP_POS_Z']
                        ] if stat >= 0
                    ])
                else:
                    total_hmean = hmean([
                        stat for stat in
                        [player.stats['R_Z'], player.stats['RBI_Z'], player.stats['HR_Z'], player.stats['SB_Z'], player.stats['Hits_Z']]
                        if stat >= 0
                    ])
                    total_pos_hmean = hmean([
                        stat for stat in
                        [player.stats['R_POS_Z'], player.stats['RBI_POS_Z'], player.stats['HR_POS_Z'], player.stats['SB_POS_Z'], player.stats['Hits_POS_Z']]
                        if stat >= 0
                    ])
                player_rollup = [player.POS, total_z, total_pos_z, total_hmean, total_pos_hmean, total_p]
                player_results.append(player_rollup)
    return player_results

In [ ]:
def simulate_draft(player_pool, db):
    draft = Draft(player_pool = player_pool, db = db, teams = 10, autodraft_level = random.randint(25,95))
    draft.run_full_autodraft()
    return draft_to_player_results(draft)

In [ ]:
all_player_results = []

for i in range(500):
    all_player_results.extend(simulate_draft(all_players, db))
    
all_player_results_df = pd.DataFrame(all_player_results, columns=['Pos','total_z','total_pos_z','total_hmean','total_pos_hmean','total_p'])

In [ ]:
all_player_results_df

In [ ]:
import numpy as np
from sklearn.linear_model import LinearRegression
X = all_player_results_df[['total_z','total_pos_z','total_hmean','total_pos_hmean']]
y = all_player_results_df['total_p']
reg = LinearRegression().fit(X, y)

In [ ]:
reg.score(X, y)

In [ ]:
reg.coef_

In [ ]:
reg.intercept_